### ISSUE TITLE
Unable to remove field lock on ULDD.X1

### ISSUE THREAD

**Issue by paulmoseley** [EncompassRest/EncompassRest#427]
According to the wiki, we should be able to unlock fields so that we can update them, but I can't get this to work on the ULDD.X1 field, which ICE decided to set as locked by default in a release early this year :(.  First thing I did was to try the approach suggested in the wiki:

CurrentLoan.Fields["ULDD.X1"].Locked = false;

That compiled cleanly, but didn't seem to work, the field remains locked after I run my code against it.  I discovered the FieldLockData object and verified that there was one field locked in the group of entities I was pulling in.  I removed the one lock from the FieldLockData list using the RemoveAt(0) command, which seemed to empty the list...but the ULDD.X1 field still shows as locked, and its value that I'm trying to update is unchanged, after I run my code against it.  I tried separating out this field unlock code from the field value update code, with a separate lock/update/unlock sequence, but that didn't help.  I _have_ verified that the field gets updated if I manually unlock the ULDD.X1 field before running my routine.  Any ideas what I might be missing here?  Does this old feature simply not work??  I realize this EncompassRest .NET bindings area is pretty dead these days, but thought I'd try entering an issue anyway, just in case...

PaulM

**Comment by ericcastelli** [comment]
Can we see a snippet of your code?

**Comment by paulmoseley** [comment]
Sure.  I'm updating fields in 3 entities, including the Uldd entity, so I start with this:

_CurrentLoan = await Repo.GetLoanDataByGuidAsync(loanGuid.LoanGuid, ["CustomField", "Uldd", "LoanProductData"]);_

Then, before attempting to update the fields in the various entities, I do this:

_// let's try unlocking the ULDD.X1 field here, before the final updates
string lockIdUnlockField = await Repo.LockLoanAsync(CurrentLoan);
CurrentLoan.Fields["ULDD.X1"].Locked = false;
//CurrentLoan.Uldd.UPBAmount
//FieldLockData fieldLockData = CurrentLoan.FieldLockData();
Log.Info($"tried to unlock field ULDD.X1 in loan {loaninfo.LoanNumber}");
Log.Info($"Number of fields locked in loan {loaninfo.LoanNumber} is {CurrentLoan.FieldLockData.Count}");
CurrentLoan.FieldLockData.RemoveAt(0);
Log.Info($"Number of fields locked in loan {loaninfo.LoanNumber} after RemoveAt(0) is {CurrentLoan.FieldLockData.Count}");
await Repo.UpdateLoanAsyncNoLock(loanGuid.LoanGuid, CurrentLoan);
await Repo.UnlockLoanAsync(CurrentLoan, lockIdUnlockField);_

My log file assures me that there aren't any field locks left in the entities after the RemoveAt(0) command (and there was only one before, presumably for the ULDD.X1 field (?)).  After which, I perform the field updates I wanted:

_CurrentLoan = await Repo.GetLoanDataByGuidAsync(loanGuid.LoanGuid, ["CustomField", "Uldd", "LoanProductData"]);
// now lock loan for final updates
string lockId = await Repo.LockLoanAsync(CurrentLoan);
// assign values from file to loan fields
UpdateLoanFields(loaninfo);
// update loan
await Repo.UpdateLoanAsyncNoLock(loanGuid.LoanGuid, CurrentLoan);
// unlock loan
await Repo.UnlockLoanAsync(CurrentLoan, lockId);
CurrentLoan = null;_

And on to the next loan.  Here's the part of the UpdateLoanFields method that updates the ULDD.X1 value:

_if (CurrentLoan.Fields["CX.INV.DEL.UPB.TYPE"].FormattedValue == "Scheduled")
{
    CurrentLoan.Fields["ULDD.X1"].Value = CurrentLoan.Fields["CX.GM.UPB"].Value;
}
else
{
    CurrentLoan.Fields["ULDD.X1"].Value = loaninfo.ActualPrincipalBalance;
}_

As I mentioned before, this code updates the ULDD.X1 value nicely if the field was manually unlocked before I started.  Let me know what you think of all this when you get a chance, and thanks! :)

PaulM

**Comment by ericcastelli** [comment]
OK. The issue is that you need to grab the loan again after you update the field lock bool.

Example:
``` C#
var loanId = "abcdefg";
var loan = new Loan(_encompassRestClient, loanId);
var loanApis = loan.LoanApis;

foreach (var lockInfo in await loanApis.GetLocksAsync())
{
    await loanApis.TryUnlockAsync(lockInfo.Id!, false);
}

var loanLockInfo = await loanApis.LockAsync(ResourceLockType.Shared);
loan.Fields["ULDD.X1"].Locked = false;
await _encompassRestClient.Loans.UpdateLoanAsync(loan);
loan = new Loan(_encompassRestClient, loanId); // This is the important line
loan.Fields["ULDD.X1"].Value = amount;
await _encompassRestClient.Loans.UpdateLoanAsync(loan);
await loanApis.UnlockAsync(loanLockInfo);
```

Please let me know if that resolves the issue you're running into.

**Comment by paulmoseley** [comment]
Great, thanks, I'll give that a try.  I think I was already trying to do that in my code, but I was unlocking the loan and re-locking it before re-getting it, maybe that makes a difference (?).  Anyway, this gives me confidence that the functionality works.  I'll work with this a bit and let you know how I do.  Thanks again! :)

**Comment by ericcastelli** [comment]
OK, right, I see that now. I guess I looked past that when I was trying to create my own test case and get it working myself. 

No problem at all. Please close the issue if it does end up working for you. :)

**Comment by paulmoseley** [comment]
OK, I finally figured this out.  It turns out the values for the fields' Locked property is exactly backwards; false means the field **IS** locked, and true means the field is **NOT** locked :(.  I stumbled across that while trying to figure out why the Locked value on ULDD.X1 was false even **before** I updated it to false.  Given this, I'm not sure if I need to do an extra update and get on the loan, but I think I'll leave the code that way since it's working. :)

Another thing I tried, at one point, was to use the version of GetLoanAsync that doesn't require the entities list, and therefore gets everything.  This call fails with an interesting error:

_Newtonsoft.Json.JsonReaderException: Unexpected character encountered while parsing value: {. Path 'elliUCDFields.leFields', line 1, position 917586._

I noticed that 'ElliUCDFields' is the next-to-last entity in the entity list, which makes me think that there's some error in that file, or maybe there's just too many entities now for this call to handle them all (?).  While I was looking at the entities list, I noticed one named 'FieldLockData' near the end, so I added that one to the list of those I wanted.  I won't open a separate issue for this GetLoanAsync problem until I need to (we're specifying the entities in all our gets now).

Anyway, I'm closing this issue now.  Thanks again, Eric, for helping me look at this. :)

PaulM
